enclave/src/main.rs

```diff
@@ -0,0 +1,160 @@
+use byteorder::{NativeEndian, WriteBytesExt};
+use chrono::prelude::*;
+use std::io::prelude::*;
+use std::{fs, path};
+
+mod bpf;
+use bpf::*;
+
+const RUNC_CHILD: &str = "runc:[2:INIT]";
+
+#[derive(thiserror::Error, Debug)]
+enum CheckBpfLsmError {
+    #[error("regex compilation error")]
+    RegexError(#[from] regex::Error),
+    
+    #[error("I/O error")]
+    IOError(#[from] std::io::Error),
+
+    #[error("BPF LSM is not enabled")]
+    BpfLsmDisabledError,
+}
+
+fn check_bpf_lsm_enabled() -> Result<(), CheckBpfLsmError> {
+    let rx = regex::Regex::new(r"bpf")?;
+    let path = path::Path::new("/sys").join("kernel").join("security").join("lsm");
+    let mut file = fs::File::open(path)?;
+    let mut content = String::new();
+    
+    file.read_to_string(&mut content)?;
+
+    if !rx.is_match(&content) {
+	return Err(CheckBpfLsmError::BpfLsmDisabledError);
+    }
+
+    Ok(())
+}
+
+#[derive(thiserror::Error, Debug)]
+enum HashError {
+    // #[error("char parsing error")]
+    // CharParseError,
+
+    #[error("could not convert the hash to a byte array")]
+    ByteWriteError(#[from] std::io::Error),
+}
+
+fn hash(s: &str) -> Result<Vec<u8>, HashError> {
+    let mut hash: u32 = 0;
+
+    for c in s.chars() {
+	let c_u32 = c as u32;
+	hash += c_u32;
+    }
+
+    let mut wtr = vec![];
+    wtr.write_u32::<NativeEndian>(hash)?;
+
+    Ok(wtr)
+}
+
+#[derive(thiserror::Error, Debug)]
+enum LoadProgramError {
+    #[error("hash error")]
+    HashError(#[from] HashError),
+
+    #[error("libbpf error")]
+    LibbpfError(#[from] libbpf_rs::Error),
+}
+
+fn init_runtimes(map: &mut libbpf_rs::Map) -> Result<(), LoadProgramError> {
+    let key = hash(RUNC_CHILD)?;
+    let val = vec![0, 0, 0, 0];
```

Comment:
If we know the size at runtime (4), this could be an array instead of a vector. Do we know the size?

Reply:
We do. I will switch than to an array then.